File: JavaScript/Plus One.js
// https://leetcode.com/problems/plus-one/

/**
 * @param {number[]} digits
 * @return {number[]}
 *
 * Given nums = [9,9], return [1,0,0]
 */
var plusOne = function(digits) {
  var i = digits.length - 1
  digits[i]++
  while (i > 0) {
    if (digits[i] != 10) {
      return digits
    }
    digits[i] = 0
    digits[i-1]++      
    i--
  }
  if (digits[0] == 10) {
    digits[0] = 1
    digits[digits.length] = 0
  }
  return digits
};

File: Go/066-plus-one/plus_one.go
package plusOne

func plusOne(digits []int) []int {
	carry := true
	for i := len(digits) - 1; carry && i >= 0; i-- {
		digits[i] = digits[i] + 1
		if digits[i] == 10 {
			digits[i] = 0
		} else {
			carry = false
		}
	}
	if carry {
		digits = append([]int{1}, digits...)
	}
	return digits
}
